wordpress-mobile/WordPress-iOS
# Auth token issues solver. [DO NOT MERGE]

Takes care points 3, 4 and 5 from [this issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/3964) reported by @astralbodies.

This should NOT be merged until [this PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/3956) by @astralbodies is merged.  The reason is that this PR is a branch from that PR.

Please keep in mind this is a hotfix, so even though some code has been moved, as little as possible of the existing logic has been really modified.  This means there's some code duplication.

**Test 1:**
- Comment out method `setUsername:` in `WPAccount`
- Login with a mixed case WP.com username name or a WP.com email address
- Create post, quit app without hitting the save button
- Reopen the app, it should ask you to re-authenticate (if you disabled the method above)
- Re-authenticate
- Make sure your post is still there

**Test 2:**
- Comment out method `setUsername:` in `WPAccount`
- Login with a mixed case WP.com username name or a WP.com email address
- Go to the site picker and add self-hosted blog
- Quit the app normally
- Reopen the app, it should ask you to re-authenticate (if you disabled the method above)
- Hit cancel instead of re-authing
- A warning will ask if you want to delete your local wp.com data
- Delete it
- Make sure the self hosted blog is still there, but that the wp.com related data has been removed (such as all wp.com blogs from the site picker).

**Test 3:**
- Re-enable method `setUsername:` in `WPAccount`
- Login with either email or mixed cap username
- Quit app, reopen
- Should log in normally, without issues

**Test 4:** 
- Comment out method `setUsername:` in `WPAccount`
- Login with a mixed case WP.com username name or a WP.com email address
- Quit the app normally
- Reopen the app, it should ask you to re-authenticate (if you disabled the method above)
- Try logging-in with a bad password.
- Make sure you can interact with the overlay that comes up, and dismiss it by pressing "OK".

**Test 5:**
- Make sure to repeat all previous tests for different iOS versions.

**Discussion:** (please take a moment to read and provide feedback)
- We don't currently have any procedure to make sure the user re-authenticates with the same user he was using previously, which seems a bit dangerous to me not to have.  This feels like something doable, but completely outside of the scope of a simple hot-fix (especially a nightly hot-fix like this one).  We could enhance `LoginViewController` with code to check that `account.userId` matches the remote info returned by the new authentication attempt.
- I believe we could improve this PR **drastically* by adding an automatic recovery attempt.  I've reviewed the keychain code we have and it supports case insensitive searches and returning multiple results.  If the auth token is `nil` we could try to (case-insensitively) search the keychain for all `account.username` and `account.email` matches, then try logging in with the returned authTokens and, on success, confirming that our locally stored `account.userId` matches the account data returned by the services (in order to make sure we're logging into the right account).  If the automatic attempt fails we can fall back into this code.
